packages/frontend/src/components/wallet-migration/modals/CleanKeysModal/CleanKeysModal.jsx

```diff
@@ -1,53 +1,255 @@
-    // 2. Ensure that the newly generated key in the RotateKeysModal is not deleted. 
-    // 3. Ensure that the funding account has Threshold * (# of FAKs to be removed) Near present in their account. 
-    // 4. Ensure that if an account has a single FAK OR is an Implicit Account with 0 Near, 
-    // 5. Create a data structure which stores what keys should be marked for deletion
-    // 6. For each account to be cleaned, show a modal where the user puts their FAK. 
+    const recoveryMethods = (await wallet.getRecoveryMethods())
+        .reduce((keys, { kind, publicKey }) => {
+            if (publicKey) {
+                keys[publicKey] = kind;
+            }
 
+            return keys;
+        }, {});
+
+    const allAccessKeys = await wallet.getAccessKeys(accountId);
+    const accessKeys = allAccessKeys
+        .filter(({ public_key }) =>
+            !publicKeyBlacklist.some((key) => key === public_key)
+            && recoveryMethods[public_key] !== 'ledger'
+        )
+        .map(({ public_key }) => ({
+            publicKey: public_key,
+            kind: recoveryMethods[public_key] || 'unknown',
+        }));
+
+    return {
+        accessKeys,
+        accountBalance,
+        accountId,
+        keyType,
+        totalAccessKeys: allAccessKeys.length,
+    };
+}
+
+async function deleteKeys({ accountId, publicKeysToDelete, wallet }) {
+    const account = await wallet.getAccount(accountId);
+    const signingPublicKey = await wallet.getPublicKey();
+
+    // TODO build batch delete transactions to delete 100 per transaction
+    const deleteKeyBatch = (keys) => Promise.all(keys.map(async (key) => {
+        await account.deleteKey(key);
+        // if the deleted key is the current signing key in the wallet, remove it from the localstorage keystore
+        if (key === signingPublicKey) {
+            await wallet.keyStore.removeKey(wallet.connection.networkId, accountId);
+        }
+    }));
+
+    const dop = 20;
+    let deleted = 0;
+    while (deleted <= publicKeysToDelete.length) {
+        const upperBound = deleted + dop;
+        await deleteKeyBatch(publicKeysToDelete.slice(deleted, upperBound));
+        deleted = upperBound;
+    }
+}
+
+const CleanKeysModal = ({ accounts, handleSetActiveView, onNext, onClose, rotatedKeys }) => {
+    const [state, localDispatch] = useImmerReducer(sequentialAccountImportReducer, {
+        accounts: []
+    });
+
+    const [loadingAccounts, setLoadingAccounts] = useState(true);
+    const [keysToRemove, setKeysToRemove] = useState({});
+    const dispatch = useDispatch();
+    const initialAccountIdOnStart = useSelector(selectAccountId);
+    const initialAccountId = useRef(initialAccountIdOnStart);
+    const [showConfirmCleanupModal, setShowConfirmCleanupModal] = useState(false);
+    const [showConfirmSeedphraseModal, setShowConfirmSeedphraseModal] = useState(false);
+    const [/*finishingSetupForCurrentAccount*/, setFinishingSetupForCurrentAccount] = useState(false);
```

Comment:
what is this for? why is it commented?

Reply:
It was preferable to keeping it in a `console.log` just to skirt the `eslint` rule about unreferenced variables. But since it never ended up being needed I can remove it.